pydm/tests/widgets/test_pushbutton.py

```diff
@@ -0,0 +1,566 @@
+        only the new value
+    channel_type : type
+        The type of the value to be sent to the channel
+    """
+
+    pydm_pushbutton = PyDMPushButton()
+    qtbot.addWidget(pydm_pushbutton)
+
+    pydm_pushbutton.showConfirmDialog = show_confirm_dialog
+    pydm_pushbutton.confirmMessage = confirm_message
+    pydm_pushbutton.relativeChange = is_value_relative
+    pydm_pushbutton.channeltype = channel_type
+
+    if press_value:
+        signals.new_value_signal[type(press_value)].connect(pydm_pushbutton.updatePressValue)
+        signals.new_value_signal[type(press_value)].emit(press_value)
+
+        if channel_type == str:
+            assert pydm_pushbutton.pressValue == str(press_value)
+
+    if updated_value:
+        signals.new_value_signal[type(updated_value)].connect(pydm_pushbutton.channelValueChanged)
+        signals.new_value_signal[type(updated_value)].emit(updated_value)
+
+        if channel_type == str:
+            assert pydm_pushbutton.pressValue == str(updated_value)
+        else:
+            assert pydm_pushbutton.channeltype == type(updated_value)
+
+    assert pydm_pushbutton.relativeChange == is_value_relative
+
+    if show_confirm_dialog:
+        monkeypatch.setattr(QMessageBox, 'exec_', lambda *args: confirm_dialog_response)
+
+    pydm_pushbutton.passwordProtected = is_password_protected
+    if is_password_protected:
+        # We assume the QInputDialog returns success. Further testing scenarios are performed at test_validate_password
+        plain_text_password = "$L4C_p4$$wd"
+        pydm_pushbutton.password = plain_text_password
+        monkeypatch.setattr(QInputDialog, 'getText', lambda *args: (plain_text_password, is_password_validated))
+
+    # Set up the conftest fixture to receive the value sent out to the channel
+    pydm_pushbutton.send_value_signal[channel_type].connect(signals.receiveValue)
+
+    # Send the value out to the channel
+    if updated_value:
+        signals.new_value_signal[type(updated_value)].connect(pydm_pushbutton.sendValue)
+        signals.new_value_signal[type(updated_value)].emit(updated_value)
+
+    if confirm_dialog_response == QMessageBox.No or not is_password_validated or None in (press_value, updated_value):
+        assert not signals.value
+    else:
+        if pydm_pushbutton.showConfirmDialog:
+            if confirm_message and len(confirm_message):
+                assert pydm_pushbutton.confirmMessage == confirm_message
+            else:
+                assert pydm_pushbutton.confirmMessage == PyDMPushButton.DEFAULT_CONFIRM_MESSAGE
+
+        if not is_value_relative or channel_type == str:
+            # Limit the decimal places since Python 2.7. provided more floating point
+            assert signals.value == pydm_pushbutton.channeltype(pydm_pushbutton.pressValue)
+        else:
+            assert signals.value == updated_value + pydm_pushbutton.channeltype(pydm_pushbutton.pressValue)
```

Comment:
This is a pretty complicated test. I think it is fine how it is for now, but in the future I would break this up into separate tests. Following the trail of nested `if` statements is difficult and ideally the testing logic simpler than what you are testing 😄